Implement a settings menu

janhohenheim opened the pull request:
![image](https://github.com/user-attachments/assets/263ae947-e290-41c7-b5aa-e7f8c584ccea)

Fixes #213 by adding a menu for volume control.

I know this was deemed out of scope a long time ago, but I want to resurrect that discussion, mainly for the sake of accessibility. I'd like to live in a world where Bevy jam games come with accessibility options (e.g. mouse acceleration, FOV, etc.) out of the box, and IMO that is only plausible if we give devs and users a settings menu for that.
The historical answer to this was to add it to the Bevy examples for users to look up, but
- a) we all know that a jam is on a tight deadline, so I don't expect users to spend extra time on this.
- b) Bevy should imo be as accessible out of the box as possible, and I want to extend that to this template.
- c) a settings menu is something that pretty much every single game in existence will profit from, even if only to allow audio control for the players. Many browsers don't even allow lowering the volume in a tab without muting it completely.

The other worry was that this would bloat up the template. To address this:

- d) the boilerplate for the settings screen is largely contained in the settings module, so it's supremely easy to rip it out if you don't want it.
- e) I intentionally wrote the code so that no new abstraction is introduced. Users don't need to know about this, they are free to entirely ignore it.

## Future work

- If this is merged, a future PR can make the menu accessible in-game through pausing. I left that task intentionally to a followup, as its implementation is largely orthogonal to this.
- We could also add new options to this menu, like screen resolution, MSAA samples, etc.

janhohenheim commented on src/audio.rs:
```diff
@@ -1,5 +1,19 @@
+    app.init_resource::<GlobalMusicVolumeScale>();
+    app.register_type::<GlobalMusicVolumeScale>();
+    app.add_systems(
+        PostUpdate,
```
Running in `PostUpdate` to guarantee this frame has the correct volume

janhohenheim commented on src/audio.rs:
```diff
@@ -33,5 +48,33 @@ pub struct Music;
+pub struct GlobalMusicVolumeScale(pub Volume);
+#[derive(Component, Debug, Reflect, Deref, DerefMut, Default)]
+#[reflect(Component)]
+pub struct UnscaledVolume(pub Volume);
```
Has no `Music` in the name because I'd reuse this for SFX

janhohenheim commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,138 @@
+    ));
+}
+
+struct PercentageSettings<E, B1, B2, M1, M2, I1, I2, C>
+where
+    E: Event,
+    B1: Bundle,
+    B2: Bundle,
+    I1: IntoObserverSystem<E, B1, M1> + Sync,
+    I2: IntoObserverSystem<E, B2, M2> + Sync,
+    C: Component,
```
I hate this

janhohenheim commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,138 @@
+struct MusicVolumeLabel;
+
+fn spawn_settings_screen(mut commands: Commands) {
+    let music_settings = PercentageSettings {
```
I'm using a struct here and not passing it as params because I think passing 4 weird params looks unreadable. Especially if you happen to forget whether `on_minus` or `on_plus` goes first.

janhohenheim commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,138 @@
+        value_marker: MusicVolumeLabel,
+        on_minus: lower_music_volume,
+        on_plus: raise_music_volume,
+        _marker: PhantomData,
```
I don't like that the `PhantomData` is constructed by hand, but `..default()` wouldn't make sense, as there is no sane default for these params, and `::new` would eliminate the positional param names

janhohenheim commented on src/theme/widget.rs:
```diff
@@ -88,3 +88,45 @@ where
 }
+
+/// A simple button with text and an action defined as an [`Observer`].
+pub fn button_small<E, B, M, I>(text: impl Into<String>, action: I) -> impl Bundle
```
Largely copy-pasted from `button`, dunno if that's considered bad form in UI code or not. I imagine this is better than introducing de-duplication functions, as you may want to fine-tune all components in this hierarchy.

will-hart replied:
To remove duplication I renamed the `button` from this PR into `button_base` which takes a size and font size argument. Then I have:

```rust
pub fn button<E, B, M, I>(text: impl Into<String>, action: I) -> impl Bundle
where
    E: Event,
    B: Bundle,
    I: IntoObserverSystem<E, B, M> + Sync,
{
    button_base(text, action, Vec2::new(200.0, 65.0), FontSizes::HEADER)
}

pub fn button_small<E, B, M, I>(text: impl Into<String>, action: I) -> impl Bundle
where
    E: Event,
    B: Bundle,
    I: IntoObserverSystem<E, B, M> + Sync,
{
    button_base(text, action, Vec2::new(180.0, 35.0), FontSizes::UI)
}

```

For my needs at least this is fine, font size and size are the main things that change for these kinds of "standard" buttons.

janhohenheim replied:
That's neat, I'll copy that design :)

benfrankel commented:
Have you seen [`GlobalVolume`](https://docs.rs/bevy/latest/bevy/audio/struct.GlobalVolume.html)?

janhohenheim commented:
@benfrankel I did not! TIL!
But that cannot distinguish between SFX and music, right?
Ofc, we could just ignore that distinction for the template.
Or add it in a different PR.

benfrankel commented:
IMO we should ignore the distinction for simplicity. We'll want a master volume slider one way or another, so doing only that for now seems good for a settings menu MVP.

janhohenheim commented:
Addressed your feedback. I also changed the volume of the files so that 100% is tolerable now. Dunno why the files got a bit bigger, but I will not fuss about a KiB.

benfrankel commented on src/theme/widget.rs:
```diff
@@ -47,8 +47,55 @@ pub fn label(text: impl Into<String>) -> impl Bundle {
+}
+
+/// A simple button with text and an action defined as an [`Observer`]. The button's components will be overridden by the components in `button_bundle`.
+pub fn button_base<E, B, M, I>(
```
I'd rather explicitly make this a non-pub helper function. Also, the `.insert(button_bundle)` overwriting trick is not needed anymore since you removed `Node` and `BorderRadius` components from the bundle anyways.

janhohenheim replied:
Yeah, but that way one could actually override something like the background color, which is neat :)
Will make it non-pub.

benfrankel replied:
YAGNI for now IMO. Could remove `BackgroundColor` from `button_base` as well.

janhohenheim replied:
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/5a9ecb876ee5934c80eb02cff86b5c26280b0427

benfrankel replied:
Oh, right, you can't just `(button(...), Node { ... })` because the actual button is a child entity. That's awkward. I thought we could remove `button_bundle` entirely as a consequence of this change. `SpawnObserver` would make this more ergonomic but alas.

janhohenheim replied:
Should I go back to the `.insert` version? IMO it's less footgunny as it cannot ever result in duplicate components being accidentally spawned

benfrankel replied:
Yeah, the `.insert` version is better. FWIW the duplicate component footgun exists any time you do `(bundle_function(), MyComponent)` :sweat_smile:

janhohenheim replied:
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/b60bea33aa657c1072d4511a6e08a9b42c2f466d

benfrankel commented on src/theme/widget.rs:
```diff
@@ -47,8 +47,55 @@ pub fn label(text: impl Into<String>) -> impl Bundle {
+where
+    E: Event,
+    B: Bundle,
+    I: IntoObserverSystem<E, B, M> + Sync,
```
```suggestion
    I: IntoObserverSystem<E, B, M>,
```
The `+ Sync` isn't required here I believe? Same goes for `fn button`.

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,93 @@
+            widget::header("Settings"),
+            (
+                Name::new("Settings"),
+                Node {
```
Maybe we should have a `widget::grid` somehow, since it seems we're now doing this in two (technically three) places. Although I'm not sure how it would support column alignment because that requires putting `justify_self` on each child node AFAIK. Ignore for this PR, but I'll give this some more thought later.

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,93 @@
+        children![
+            widget::header("Settings"),
+            (
+                Name::new("Settings"),
```
```suggestion
                Name::new("Settings Grid"),
```

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,93 @@
+                            ..default()
+                        }
+                    ),
+                    (
+                        Node {
+                            justify_self: JustifySelf::Start,
+                            ..default()
+                        },
+                        children![
+                            widget::button_small("-", lower_volume),
+                            (
+                                Node {
+                                    padding: UiRect::horizontal(Val::Px(10.0)),
+                                    justify_content: JustifyContent::Center,
+                                    ..default()
+                                },
+                                children![(widget::label(""), GlobalVolumeLabel)],
+                            ),
+                            widget::button_small("+", raise_volume),
+                        ],
+                    ),
```
Pull this section out to a `fn volume_widget() -> impl Bundle` IMO.

janhohenheim replied:
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/f049cd1e678d85dcd68fcfe9e936ca04e1649519

benfrankel replied:
Oh I didn't mean the whole grid. Suppose a user of this template wants to add another settings widget to the grid for resolution or whatever. It'd be cleaner to encapsulate each widget as a bundle function and add that function as another child in the settings grid.

janhohenheim replied:
Makes sense :)
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/9cd8ebf8ebcd2706c2bd8c37de80c10f1a361353

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,93 @@
+    );
+}
+
+#[derive(Component)]
+struct GlobalVolumeLabel;
```
Move this down to right above `fn update_volume_label`. Also derive `Reflect` and reflect `Component`, and add `app.register_type::<GlobalVolumeLabel>()`.

janhohenheim replied:
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/93a7716ce668cdb3d40ed6ca286111f21d8fb8d6

SecretPocketCat commented on src/audio.rs:
```diff
@@ -34,31 +20,3 @@ pub struct SoundEffect;
-        let sink = sink.bypass_change_detection();
-        sink.set_volume(music_volume_factor.0 * unscaled_volume.0);
-    }
-}
```
Smt. like this might still be needed based on the note from [GlobalVolume](https://docs.rs/bevy/latest/bevy/audio/struct.GlobalVolume.html).

> Note: Changing [GlobalVolume](https://docs.rs/bevy/latest/bevy/prelude/struct.GlobalVolume.html) does not affect already playing audio.

janhohenheim replied:
heck

janhohenheim replied:
Opened an issue for this: https://github.com/bevyengine/bevy/issues/18952

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,99 @@
+const MAX_VOLUME: f32 = 3.0;
+
+pub(super) fn plugin(app: &mut App) {
+    app.register_type::<GlobalVolumeLabel>();
+    app.add_systems(OnEnter(Screen::Settings), spawn_settings_screen);
+    app.add_systems(
+        Update,
+        update_volume_label.run_if(in_state(Screen::Settings)),
+    );
```
```suggestion
    app.add_systems(OnEnter(Screen::Settings), spawn_settings_screen);
    
    app.register_type::<GlobalVolumeLabel>();
    app.add_systems(
        Update,
        update_volume_label.run_if(in_state(Screen::Settings)),
    );
```
Nit

benfrankel commented on src/screens/settings.rs:
```diff
@@ -0,0 +1,100 @@
+
+use crate::{screens::Screen, theme::prelude::*};
+
+const MIN_VOLUME: f32 = 0.0;
+const MAX_VOLUME: f32 = 3.0;
+
```
I'd move these to just below `fn volume_widget()`. Not blocking.

janhohenheim replied:
https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/f54a34db650721627b1d2e40219f5e01fba36d5c

benfrankel commented on src/theme/widget.rs:
```diff
@@ -59,19 +106,11 @@ where
         Name::new("Button"),
         Node::default(),
-        Children::spawn(SpawnWith(|parent: &mut ChildSpawner| {
+        Children::spawn(SpawnWith(move |parent: &mut ChildSpawner| {
```
```suggestion
        Children::spawn(SpawnWith(|parent: &mut ChildSpawner| {
```
Is the `move` still needed?

janhohenheim commented on src/audio.rs:
```diff
@@ -12,9 +12,6 @@ pub fn music(handle: Handle<AudioSource>) -> impl Bundle {
 
-/// An organizational marker component that should be added to a spawned [`AudioPlayer`] if it's in the
-/// general "sound effect" category (e.g. footsteps, the sound of a magic spell, a door opening).
-///
```
Whoops https://github.com/TheBevyFlock/bevy_new_2d/pull/374/commits/99a9119af3ff484a56e419f95c67fb14f00094f8

janhohenheim commented on src/lib.rs:
```diff
@@ -43,12 +39,6 @@ impl Plugin for AppPlugin {
                     }
                     .into(),
                     ..default()
-                })
-                .set(AudioPlugin {
-                    global_volume: GlobalVolume {
-                        volume: Volume::Linear(0.3),
-                    },
-                    ..default()
```
I removed it because it seems weird for the player to have the default audio be at 80%, and I didn't want to add the extra boilerplate of doing math with `const DEFAULT_GLOBAL_VOLUME: f32 = 0.3`

benfrankel reviewed (approved):
Looks good enough to merge now :tada:

We can continue to refine the settings menu in future PRs, maybe adding a pause menu that can navigate to settings as well.
